compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -53,14 +57,48 @@ pub trait NormalizeExt<'tcx> {
     /// This normalization should be used when the type contains inference variables or the
     /// projection may be fallible.
     fn normalize<T: TypeFoldable<TyCtxt<'tcx>>>(&self, t: T) -> InferOk<'tcx, T>;
+
+    /// Deeply normalizes `value`, replacing all aliases which can by normalized in
+    /// the current environment. Unlike other normalization routines, this errors
+    /// in case normalization fails or is ambiguous.
+    ///
+    /// In the old solver this simply uses `normalize` and errors in
+    /// case of ambiguity. The new solver only normalizes in this function and
+    /// `normalize` is a noop.
+    ///
+    /// This only normalize opaque types with `Reveal::All`.
+    fn deeply_normalize<T: TypeFoldable<TyCtxt<'tcx>>>(
+        self,
+        value: T,
+    ) -> Result<T, Vec<FulfillmentError<'tcx>>>;
 }
 
 impl<'tcx> NormalizeExt<'tcx> for At<'_, 'tcx> {
     fn normalize<T: TypeFoldable<TyCtxt<'tcx>>>(&self, value: T) -> InferOk<'tcx, T> {
-        let mut selcx = SelectionContext::new(self.infcx);
-        let Normalized { value, obligations } =
-            normalize_with_depth(&mut selcx, self.param_env, self.cause.clone(), 0, value);
-        InferOk { value, obligations }
+        if self.infcx.next_trait_solver() {
+            InferOk { value, obligations: Vec::new() }
+        } else {
+            let mut selcx = SelectionContext::new(self.infcx);
+            let Normalized { value, obligations } =
+                normalize_with_depth(&mut selcx, self.param_env, self.cause.clone(), 0, value);
+            InferOk { value, obligations }
+        }
+    }
+
+    fn deeply_normalize<T: TypeFoldable<TyCtxt<'tcx>>>(
```

Comment:
Naming thought: Should this perhaps just be called [`fully_normalize`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_trait_selection/traits/fn.fully_normalize.html)? since it basically does what that does.

Reply:
And I think calls to `fully_normalize` should get migrated to this one, probably.

Reply:
added comments for that. didn't rename this yet to not make the confusion worse